File: hallucinate_app/node/examples/error_reporting_example.js
#!/usr/bin/env node
/**
 * Example: Auto Error Reporting (JavaScript)
 * 
 * Demonstrates how the auto error reporting system works in JavaScript.
 * Run this with GITHUB_ISSUE_DRY_RUN=true to test without creating issues.
 */

import {
  GitHubIssueReporter,
  IssueReportConfig,
  getReporter,
  reportError,
  reportException,
  ErrorLevel,
  ErrorSource
} from '../../node/github_issue_reporter.js';

async function main() {
  console.log('='.repeat(60));
  console.log('Auto Error Reporting Example (JavaScript)');
  console.log('='.repeat(60));
  console.log();
  
  // Example 1: Direct use of the GitHub reporter
  console.log('Example 1: Direct GitHub reporter usage');
  console.log('-'.repeat(60));
  
  // Configure for dry run mode
  const config = new IssueReportConfig({
    enabled: true,
    githubToken: process.env.GITHUB_TOKEN || 'test_token',
    repository: process.env.GITHUB_REPOSITORY || 'owner/repo',
    minErrorLevel: ErrorLevel.ERROR,
    dryRun: true  // Safe for testing
  });
  
  const reporter = new GitHubIssueReporter(config);
  
  // Create a sample error
  const error = {
    id: 'example-js-001',
    timestamp: new Date().toISOString(),
    level: ErrorLevel.ERROR,
    source: ErrorSource.JAVASCRIPT,
    component: 'ExampleComponent',
    operation: 'example_operation',
    message: 'This is an example error message from JavaScript',
    stackTrace: 'Error: Example error\n    at example.js:10:15\n    at main (example.js:5:3)',
    details: { context: 'This happened during testing' },
    metadata: {
      nodeVersion: process.version,
      platform: process.platform,
      arch: process.arch
    },
    tags: ['example', 'test', 'documentation', 'javascript']
  };
  
  // Report the error
  const result = await reporter.createIssue(error);
  
  if (result) {
    console.log(`✓ Issue would be created: ${result}`);
  } else {
    console.log('✗ Issue was not created (check configuration)');
  }
  
  console.log();
  console.log('Statistics:', reporter.getStats());
  console.log();
  
  // Example 2: Using reportException helper
  console.log('Example 2: Using reportException helper');
  console.log('-'.repeat(60));
  
  try {
    // This will throw an error
    throw new Error('This is an example exception from JavaScript');
  } catch (err) {
    // Report the exception
    const issueUrl = await reportException(err, {
      component: 'ExampleComponent',
      operation: 'example_operation',
      source: ErrorSource.NODEJS,
      level: ErrorLevel.ERROR,
      details: { context: 'Testing reportException helper' }
    });
    
    if (issueUrl) {
      console.log(`✓ Exception reported: ${issueUrl}`);
    } else {
      console.log('✓ Exception processed (dry run mode - no actual issue created)');
    }
  }
  
  console.log();
  
  // Example 3: Demonstrating duplicate detection
  console.log('Example 3: Duplicate detection');
  console.log('-'.repeat(60));
  
  const duplicateError = {
    id: 'example-js-002',
    timestamp: new Date().toISOString(),
    level: ErrorLevel.ERROR,
    source: ErrorSource.JAVASCRIPT,
    component: 'ExampleComponent',
    operation: 'example_operation',
    message: 'This is an example error message from JavaScript',  // Same message as before
    stackTrace: 'Error: Example error\n    at example.js:10:15',
    tags: ['example', 'test']
  };
  
  const result1 = await reporter.createIssue(duplicateError);
  console.log(`First attempt: ${result1 ? 'Created' : 'Not created'}`);
  
  const result2 = await reporter.createIssue(duplicateError);
  console.log(`Second attempt (duplicate): ${result2 ? 'Created' : 'Not created (duplicate detected)'}`);
  
  console.log();
  console.log('Updated statistics:', reporter.getStats());
  
  console.log();
  console.log('='.repeat(60));
  console.log('Examples completed!');
  console.log();
  console.log('To enable actual GitHub issue creation:');
  console.log('1. Set GITHUB_TOKEN environment variable');
  console.log('2. Set GITHUB_REPOSITORY environment variable');
  console.log('3. Set GITHUB_ISSUE_REPORTER_ENABLED=true');
  console.log('4. Remove GITHUB_ISSUE_DRY_RUN or set it to false');
  console.log('='.repeat(60));
}

// Run the examples
main().catch(err => {
  console.error('Error running examples:', err);
  process.exit(1);
});

File: hallucinate_app/python/hallucinate_app/examples/error_reporting_example.py
#!/usr/bin/env python3
"""
Example: Auto Error Reporting

Demonstrates how the auto error reporting system works.
Run this with GITHUB_ISSUE_DRY_RUN=true to test without creating issues.
"""

import os
import sys
import asyncio

# Set up the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from hallucinate_app.github_issue_reporter import (
    GitHubIssueReporter,
    IssueReportConfig,
    report_error
)
from hallucinate_app.error_monitor import (
    ErrorMonitor,
    ErrorData,
    ErrorLevel,
    ErrorSource
)


async def main():
    """Run examples"""
    
    print("=" * 60)
    print("Auto Error Reporting Example")
    print("=" * 60)
    print()
    
    # Example 1: Direct use of the GitHub reporter
    print("Example 1: Direct GitHub reporter usage")
    print("-" * 60)
    
    # Configure for dry run mode
    config = IssueReportConfig(
        enabled=True,
        github_token=os.getenv('GITHUB_TOKEN', 'test_token'),
        repository=os.getenv('GITHUB_REPOSITORY', 'owner/repo'),
        min_error_level=ErrorLevel.ERROR,
        dry_run=True  # Safe for testing
    )
    
    reporter = GitHubIssueReporter(config)
    
    # Create a sample error
    error = ErrorData(
        id='example-001',
        timestamp='2024-01-15T10:30:00.000Z',
        level=ErrorLevel.ERROR,
        source=ErrorSource.PYTHON,
        component='ExampleComponent',
        operation='example_operation',
        message='This is an example error message',
        stack_trace='Traceback (most recent call last):\n  File "example.py", line 10, in example_operation\n    raise ValueError("Example error")\nValueError: Example error',
        details={'context': 'This happened during testing'},
        metadata={'python_version': sys.version, 'platform': sys.platform},
        tags=['example', 'test', 'documentation']
    )
    
    # Report the error
    result = reporter.create_issue(error)
    
    if result:
        print(f"✓ Issue would be created: {result}")
    else:
        print("✗ Issue was not created (check configuration)")
    
    print()
    print(f"Statistics: {reporter.get_stats()}")
    print()
    
    # Example 2: Using the ErrorMonitor integration
    print("Example 2: ErrorMonitor integration")
    print("-" * 60)
    
    # Configure error monitor with GitHub reporting
    monitor_config = {
        'enable_github_reporting': True,
        'github_config': {
            'enabled': True,
            'github_token': os.getenv('GITHUB_TOKEN', 'test_token'),
            'repository': os.getenv('GITHUB_REPOSITORY', 'owner/repo'),
            'min_error_level': ErrorLevel.ERROR,
            'dry_run': True
        }
    }
    
    monitor = ErrorMonitor(config=monitor_config)
    await monitor.start()
    
    # Simulate an exception
    try:
        # This will raise an exception
        raise ValueError("This is an example exception from the error monitor")
    except Exception as e:
        # Add to error monitor (will automatically report to GitHub)
        error_id = await monitor.add_exception(
            e,
            component='ExampleComponent',
            operation='example_operation',
            source=ErrorSource.PYTHON,
            level=ErrorLevel.ERROR,
            details={'context': 'Testing error monitor integration'}
        )
        
        print(f"✓ Exception added to error monitor: {error_id}")
        
        # Give the monitor time to process
        await asyncio.sleep(2)
        
        # Get the error details
        error_data = await monitor.get_error(error_id)
        if error_data and 'github_issue_url' in error_data.metadata:
            print(f"✓ GitHub issue URL: {error_data.metadata['github_issue_url']}")
        else:
            print("✓ Error processed (dry run mode - no actual issue created)")
    
    await monitor.stop()
    
    print()
    print("=" * 60)
    print("Examples completed!")
    print()
    print("To enable actual GitHub issue creation:")
    print("1. Set GITHUB_TOKEN environment variable")
    print("2. Set GITHUB_REPOSITORY environment variable")
    print("3. Set GITHUB_ISSUE_REPORTER_ENABLED=true")
    print("4. Remove GITHUB_ISSUE_DRY_RUN or set it to false")
    print("=" * 60)


if __name__ == '__main__':
    asyncio.run(main())
